### PR TITLE
Restructure `feos-core`

### PR DESCRIPTION
**Pull request by g-bauer**
This PR restructures some elements of `feos-core` that make the code easier and a bit less opinionated. 

- Removal of the methods that return trait objects from `Residual` and `IdealGas`: the interface for the Helmholtz energy calculation no longer distinguishes between different contributions. 
- Consequently, we get rid of the traits and constructs to create object-safe traits.
- Since the interface for the Helmholtz energy is less rigid we can think about efficiently calculating properties that are used across different contributions or even caching.

### CODE REVIEW

File: feos-core/src/equation_of_state/ideal_gas.rs [hunk 1]
```diff
@@ -2,24 +2,21 @@ use super::Components;
 use crate::StateHD;
 use ndarray::Array1;
 use num_dual::DualNum;
-use num_dual::*;
-use std::fmt;
 
 /// Ideal gas Helmholtz energy contribution.
 pub trait IdealGas: Components + Sync + Send {
```
[the remark this answers is not in the document]
Comment (prehner) [reply feos-core/src/equation_of_state/ideal_gas.rs:7]: That was actually a bad idea :(

File: src/gc_pcsaft/eos/mod.rs [hunk 2]
```diff
@@ -98,8 +109,35 @@ impl Residual for GcPcSaft {
...
+    ) -> Vec<(String, D)> {
+        let mut v = Vec::with_capacity(7);
+        let d = self.parameters.hs_diameter(state.temperature);
+
+        v.push((
+            "Hard Sphere".to_string(),
+            self.hard_sphere.helmholtz_energy(state),
+        ));
+        v.push((
+            "Hard Sphere".to_string(),
```
Comment (prehner) [inline src/gc_pcsaft/eos/mod.rs:124]:
```suggestion
            self.hard_chain.to_string(),
```

File: src/pcsaft/eos/dispersion.rs [hunk 3]
```diff
@@ -65,28 +65,36 @@ pub struct Dispersion {
...
+        diameter: &Array1<D>,
+    ) -> D {
         // auxiliary variables
         let n = self.parameters.m.len();
         let p = &self.parameters;
         let rho = &state.partial_density;
 
         // temperature dependent segment radius
-        let r = p.hs_diameter(state.temperature) * 0.5;
+        // let r = p.hs_diameter(state.temperature) * 0.5;
```
Comment (prehner) [inline src/pcsaft/eos/dispersion.rs:80]:
```suggestion
```

File: src/pcsaft/eos/polar.rs [hunk 4]
```diff
@@ -180,7 +184,8 @@ impl<D: DualNum<f64> + Copy> HelmholtzEnergyDual<D> for Dipole {
             .collect();
 
         let rho = &state.partial_density;
-        let r = p.hs_diameter(state.temperature) * 0.5;
+        // let r = p.hs_diameter(state.temperature) * 0.5;
```
Comment (prehner) [inline src/pcsaft/eos/polar.rs:187]:
```suggestion
```

File: src/pcsaft/eos/polar.rs [hunk 5]
```diff
@@ -257,7 +267,8 @@ impl<D: DualNum<f64> + Copy> HelmholtzEnergyDual<D> for Quadrupole {
...
```
Comment (prehner) [inline src/pcsaft/eos/polar.rs:270]:
```suggestion
```

File: src/pcsaft/eos/polar.rs [hunk 6]
```diff
@@ -348,7 +364,8 @@ impl<D: DualNum<f64> + Copy> HelmholtzEnergyDual<D> for DipoleQuadrupole {
...
```
Comment (prehner) [inline src/pcsaft/eos/polar.rs:367]:
```suggestion
```

File: src/saftvrqmie/eos/dispersion.rs [hunk 7]
```diff
@@ -90,31 +92,40 @@ pub struct Dispersion {
...
         // auxiliary variables
         let n = self.parameters.m.len();
         let p = &self.parameters;
         let rho = &state.partial_density;
+
         // temperature dependent segment radius
-        let s_eff_ij = Array2::from_shape_fn((n, n), |(i, j)| -> D {
-            p.calc_sigma_eff_ij(i, j, state.temperature)
-        });
+        // let s_eff_ij = Array2::from_shape_fn((n, n), |(i, j)| -> D {
```
Comment (prehner) [inline src/saftvrqmie/eos/dispersion.rs:107]: remove comments

File: examples/gc_pcsaft_functional.ipynb
[the remark this answers is not in the document]
Comment (g-bauer) [reply examples/gc_pcsaft_functional.ipynb:1]: Good catch. Removed them.

File: Cargo.toml [hunk 8]
```diff
@@ -53,6 +53,8 @@ criterion = "0.5"
 [profile.release-lto]
 inherits = "release"
 lto = true
+codegen-units = 1
```
[the remark this answers is not in the document]
Comment (g-bauer) [reply Cargo.toml:56]: Yes, I'd keep it.

File: feos-core/src/density_iteration.rs [hunk 9]
```diff
@@ -135,7 +135,7 @@ pub fn density_iteration<E: Residual>(
     }
 }
 
-fn pressure_spinodal<E: Residual>(
+pub(crate) fn pressure_spinodal<E: Residual>(
```
[the remark this answers is not in the document]
Comment (g-bauer) [reply feos-core/src/density_iteration.rs:138]: Good catch! Removed pub.

File: feos-core/src/equation_of_state/ideal_gas.rs [hunk 10]
```diff
@@ -2,24 +2,20 @@ use super::Components;
...
-    // Return a reference to the implementation of the de Broglie wavelength.
-    fn ideal_gas_model(&self) -> &dyn DeBroglieWavelength;
+    fn ln_lambda3<D: DualNum<f64> + Copy>(&self, temperature: D) -> Array1<D>;
+
+    fn ideal_gas_model(&self) -> String;
```
[the remark this answers is not in the document]
Comment (g-bauer) [reply feos-core/src/equation_of_state/ideal_gas.rs:10]: Done.

**Pull request merged**